Add incremental snapshot utils

brooksprumo opened the pull request:
This commit adds high-level functions (`bank_to_incremental_snapshot()` and `bank_from_incremental_snapshot()`), and additionally
refactors snapshot_utils and serde_snapshot to use a common code path for
snapshots.  New tests were added to snapshot_utils for incremental snapshots.

This PR grabs bits from PR #17875, and incorporated feedback from that PR review. This PR is another in the pursuit of Incremental Snapshots (#17088), and this PR finally/actually adds the ability to both load-from and store-to an incremental snapshot. Yay!


This PR is quite big. A lot of the deserialization code was touched to support incremental snapshots and avoid code duplication. The diff may be easier to read when ignoring whitespace.

The next PR in this series will add new integration-style tests to `core/tests/`, and will address new issues those tests uncover (likely `clean_accounts()`).

#### Additional Bits

Care has been taken to rename (almost) everything around snapshots.
- Wherever specific, all variables and functions will be `full_` or `incremental_` unless they work for both.
- The term "snapshot" was overloaded in the code; sometimes called a bank snapshot, sometimes a slot snapshot, and sometimes referring to the snapshot package/snapshot archive. Outside of Bank, it is now a "bank snapshot".


## Outstanding Questions

- [x] **Naming: Should it be "snapshot" and "incremental snapshot" , or "full snapshot" and "incremental snapshot"** There's quite a bit of existing non-code that uses just "snapshot"; should that be updated too? [More discussion here](https://github.com/solana-labs/solana/pull/18504#discussion_r668915990)
    - resolved by renaming
- [x] **Assert vs error when storages overlap.** [Thread 1 here](https://github.com/solana-labs/solana/pull/18504#discussion_r668281871), [thread 2 here](https://github.com/solana-labs/solana/pull/18504#discussion_r668284384), and [comment 3 here](https://github.com/solana-labs/solana/pull/18504#issuecomment-883501258)
    - resolved by error-ing instead of panic-ing
- [x] **Should I remove `deserialize_snapshot_data_file_capped()`?** [Thread here](https://github.com/solana-labs/solana/pull/18504#discussion_r668267484)
- [x] **Should I create another struct for common parameters?** [Thread here](https://github.com/solana-labs/solana/pull/18504#discussion_r668294748)

## New Issues Created

- #18639
- #18641
- #18813 
- #18815 
- #18841

brooksprumo commented on runtime/src/serde_snapshot.rs:
```diff
@@ -126,6 +141,7 @@ where
         .deserialize_from::<R, T>(reader)
 }
 
+#[cfg(test)]
 #[allow(clippy::too_many_arguments)]
 pub(crate) fn bank_from_stream<R>(
```
This function is now only used in serde_snapshots/tests.rs. I could remove it and update the test, but figured it was fine to leave it. Let me know if it should be removed instead.

carllin replied:
yeah looks like this is just a convenience function around passing a `SnapshotStreams` at this point. I think this can be removed and the functionality just moved inside the one test

brooksprumo replied:
OK will do!

brooksprumo replied:
Done!

brooksprumo commented on runtime/src/snapshot_utils.rs:
```diff
@@ -814,23 +1387,66 @@ where
+            .unwrap_or_else(|err| info!("Failed to remove old snapshot: {}", err));
     }
+
+    // Only keep incremental snapshots for the latest full snapshot
+    // bprumo TODO: As an option to further reduce the number of incremental snapshots, only a
+    // subset of the incremental snapshots for the lastest full snapshot could be kept.  This could
+    // reuse maximum_snapshots_to_retain, or use a new field just for incremental snapshots.
```
I plan to address this TODO in a subsequent PR when adding integration tests for incremental snapshots.

lijunwangs replied:
I think at interface level maximum_snapshots_to_retain is good enough. Snapshot is a snapshot. Incremental or full is implementation detail. for efficiency. The user may be advised that with incremental feature they can afford to keep more snapshots around for the same storage overhead.

brooksprumo replied:
I've created a new issue to track this TODO, #18639. I think this may need some additional configuration options. Stephen has mentioned wanting to keep old (full) snapshots around longer for replay purposes, and I could see similar configurations requested for incremental snapshots.

brooksprumo commented on runtime/src/serde_snapshot.rs:
```diff
@@ -146,18 +200,33 @@ where
-
+            let (full_snapshot_bank_fields, full_snapshot_accounts_db_fields) =
+                $x::deserialize_bank_fields(snapshot_streams.full_snapshot_stream)?;
+            let (incremental_snapshot_bank_fields, incremental_snapshot_accounts_db_fields) =
+                if let Some(ref mut incremental_snapshot_stream) =
+                    snapshot_streams.incremental_snapshot_stream
+                {
+                    let (bank_fields, accounts_db_fields) =
+                        $x::deserialize_bank_fields(incremental_snapshot_stream)?;
+                    (Some(bank_fields), Some(accounts_db_fields))
+                } else {
+                    (None, None)
+                };
```
I use this pattern a number of times to handle `Option<>` on incremental snapshot stuffs. While this way works, it feels very redundant and verbose. Is there a better/clearer way to do this?

carllin replied:
This seems to make sense to me, the `?` in `$x::deserialize_bank_fields(incremental_snapshot_stream)?;` makes it hard to use combinators

brooksprumo replied:
Yeah... I was hoping there was some other Rust magic out there :)

carllin commented on runtime/src/serde_snapshot.rs:
```diff
@@ -321,19 +390,54 @@ where
+        full_snapshot_bank_hash_info,
+    ) = snapshot_accounts_db_fields.full_snapshot_accounts_db_fields;
+
+    let incremental_snapshot_version = snapshot_accounts_db_fields
+        .incremental_snapshot_accounts_db_fields
+        .as_ref()
+        .map(|fields| fields.1);
+    let incremental_snapshot_slot = snapshot_accounts_db_fields
+        .incremental_snapshot_accounts_db_fields
+        .as_ref()
+        .map(|fields| fields.2);
+    let incremental_snapshot_bank_hash_info = snapshot_accounts_db_fields
+        .incremental_snapshot_accounts_db_fields
+        .as_ref()
+        .map(|fields| fields.3.clone());
```
How about 

```
let (snapshot_version, snapshot_slot, snapshot_bank_hash_info) = snapshot_accounts_db_fields
        .incremental_snapshot_accounts_db_fields
        .as_ref()
        .map(
            |AccountsDbFields(
                _,
                incremental_snapshot_version,
                incremental_snapshot_slot,
                incremental_snapshot_bank_hash_info,
            )| {
                (
                    *incremental_snapshot_version,
                    *incremental_snapshot_slot,
                    incremental_snapshot_bank_hash_info.clone(),
                )
            },
        )
        .unwrap_or((
            full_snapshot_version,
            full_snapshot_slot,
            full_snapshot_bank_hash_info,
        ));
```

and removing the `unwrap_or` for each of the references below

brooksprumo replied:
Love it! Done.

carllin commented on runtime/src/serde_snapshot.rs:
```diff
@@ -367,11 +471,10 @@ where
+    );
 
     // Process deserialized data, set necessary fields in self
     let max_id: usize = *storage
```
Is it guaranteed that between the full snapshot generated on one node, an the incremental snapshot generated/fetched from another node, that *none* the storage entries will have overlapping ids?

brooksprumo replied:
Right, yeah, I remember your comment on the other PR about the AppendVec IDs. I would assume there's no guarantee that different nodes will have the same IDs, therefore there may be overlapping IDs. I think there are a few options for this PR:

1. Re-map duplicate IDs before reconstructing the storage map.
2. Leave the implementation as-is, and then address the overlapping IDs in a later PR.
3. Prevent duplicate IDs by requiring incremental snapshots come from the same node that provides the full snapshot.

Originally I was thinking (3) would be the way to go. Maybe this is too restrictive though? How often does a node download a full snapshot? I would guess not very often, right? After a node has joined, it creates its own full snapshots and incremental snapshots for future restarts. And I think on the gossip/RPC side this is the plan; i.e. when a node asks for a snapshot and downloads it, it will be given a full snapshot and an incremental snapshot.

For (2), I think this is also my vote, regardless of (3). I tried to make this PR as small as possible, and just have the necessary bits to create and rebuild from incremental snapshots. So then (1) could be added later as necessary, when it becomes a requirement to enable getting full snapshots and incremental snapshots from different nodes.

Is that OK for this PR?

carllin replied:
> Maybe this is too restrictive though?

Agreed, I think this is too restrictive.

> For (2), I think this is also my vote, regardless of (3). I tried to make this PR as small as possible, and just have the necessary bits to create and rebuild from incremental snapshots. So then (1) could be added later as necessary, when it becomes a requirement to enable getting full snapshots and incremental snapshots from different nodes.

Yeah that makes sense, as long as that code makes it in before the integration with gossip it should be fine, maybe an issue to track it?

brooksprumo replied:
OK, new issue has been created to track this: #18641.

carllin replied:
thanks!

carllin commented on runtime/src/snapshot_utils.rs:
```diff
@@ -53,18 +55,42 @@ pub struct SnapshotArchiveInfo {
+    pub base_slot: Slot,
+
+    /// Slot that the incremental snapshot was made
+    pub slot: Slot,
+
+    /// Hash of the accounts at this slot
+    pub hash: Hash,
+
+    /// Archive format for the incremental snapshot file
+    pub archive_format: ArchiveFormat,
```
does it make sense to just have `IncrementalSnapshotArchiveInfo` hold a `SnapshotArchiveInfo` and a `base_slot`

lijunwangs replied:
I second that -- it does look a lot of fields have the same names and have same meanings.

brooksprumo replied:
Love it. Will do!

brooksprumo replied:
Done.

carllin commented on runtime/src/snapshot_utils.rs:
```diff
@@ -649,64 +896,280 @@ pub fn bank_from_snapshot_archive<P>(
+    frozen_account_pubkeys: &[Pubkey],
+    snapshot_path: &Path,
+    full_snapshot_archive_path: P,
+    incremental_snapshot_archive_path: Q,
```
If we just passed this as an option to the existing `bank_from_snapshot_archive` could we avoid this function?

brooksprumo replied:
Yep, can have just the one top-level function and not add a new one. Will do!

brooksprumo replied:
Done!

carllin commented on runtime/src/snapshot_utils.rs:
```diff
@@ -536,6 +677,112 @@ where
+where
+    P: AsRef<Path>,
+{
+    let consumed_size = file_stream.seek(SeekFrom::Current(0))?;
```
I think this is equivalent to `file_stream.stream_position()`

brooksprumo replied:
Neat! Yeah, the actual code wasn't changed from the original implementation, just moved. I'll make this change though, since I agree it is simpler.

brooksprumo replied:
Done.

carllin commented on runtime/src/snapshot_utils.rs:
```diff
@@ -536,6 +677,112 @@ where
+
+/// After running the deserializer function, perform common checks to ensure the snapshot archive
+/// files were consumed correctly.
+fn check_deserialize_snapshot_data_file<P>(
```
nit: `check_deserialize_snapshot_data_file -> check_deserialize_file_consumed/check_file_consumed`

brooksprumo replied:
Done.

carllin commented on runtime/src/snapshot_utils.rs:
```diff
@@ -536,6 +677,112 @@ where
     Ok(ret)
 }
 
+fn deserialize_snapshot_data_files_capped<F, T>(
```
Nice, can the previous version `deserialize_snapshot_data_file_capped` be deprecated/removed now in favor of this version?

brooksprumo replied:
Yes, that can be done. I was wondering about that myself. This function is also called when deserializing the status cache[1], and that's just a single file, so I left this function here for that. I could change the status cache code to call this version, but I thought that looked clunky and didn't communicate intent well. Whatcha think?

[1] https://github.com/brooksprumo/solana/blob/4f45db66c833ab1828c668c84de05b55e26a83e8/runtime/src/snapshot_utils.rs#L1634

brooksprumo replied:
Oh, sorry, I think I misunderstood ^^^

Yes, the `_capped()` versions can be collapsed into just one. That's the intent, yes?

carllin replied:
> Yes, the _capped() versions can be collapsed into just one. That's the intent, yes?

yeah it seems like one function should be enough to handle either just a full snapshot or a full + incremental snapshot

brooksprumo replied:
Gotcha. Done!

carllin commented on runtime/src/snapshot_utils.rs:
```diff
@@ -932,46 +1551,86 @@ fn verify_snapshot_version_and_folder(
     frozen_account_pubkeys: &[Pubkey],
-    unpacked_snapshots_dir: &Path,
+    full_snapshot_unpacked_snapshots_dir: P,
+    incremental_snapshot_unpacked_snapshots_dir: Option<Q>,
```
Might be a good idea to fold `incremental_snapshot_unpacked_snapshots_dir` and `incremental_snapshot_version` into a tuple under the same `Option` here since they are both Some or None, maybe a struct `UnpackedSnapshotDir`? (maybe could use this for the `full_snapshot_unpacked_snapshots_dir + full_snapshot_version` as well

brooksprumo replied:
I like it. Will do!

brooksprumo replied:
Done!

carllin commented on runtime/src/snapshot_utils.rs:
```diff
@@ -649,64 +896,280 @@ pub fn bank_from_snapshot_archive<P>(
+    account_paths: &[PathBuf],
+    archive_format: ArchiveFormat,
+    parallel_divisions: usize,
+) -> Result<(TempDir, PathBuf, UnpackedAppendVecMap, String, Measure)>
```
This is a rather large tuple that could deserve a struct. Might be good to also include the struct described here: https://github.com/solana-labs/solana/pull/18504/files#r668273623   as a sub-struct of that struct

brooksprumo replied:
I agree! Will do.

brooksprumo replied:
Done!

carllin commented on runtime/src/snapshot_utils.rs:
```diff
@@ -932,46 +1551,86 @@ fn verify_snapshot_version_and_folder(
     info!(
-        "Loading bank from {}",
-        &root_paths.snapshot_file_path.display()
+        "Loading bank from {} and {:?}",
```
nit: `Loading bank from full snapshot: {} and incremental snapshot: {:?}`

brooksprumo replied:
Same naming questions from here [1] on how to name snapshots.

[1] https://github.com/solana-labs/solana/pull/18504#discussion_r668915990

brooksprumo replied:
Done!

carllin commented on runtime/src/snapshot_utils.rs:
```diff
@@ -476,6 +603,20 @@ where
 
+fn deserialize_snapshot_data_files<F, T>(
+    snapshot_root_paths: &SnapshotRootPaths,
+    deserializer: F,
```
nit: throughout the codebase we've been using `deserializer: impl FnOnce...` instead of the `where` clause below. Might be good to keep that consistent across all the functions here.

brooksprumo replied:
Ah OK. I was following the existing impl of the file; I'll happily make these change though!

brooksprumo replied:
Done.

carllin commented on runtime/src/serde_snapshot.rs:
```diff
@@ -321,19 +351,64 @@ where
+                .collect::<HashMap<_, _>>()
+        });
+
+    assert!(incremental_snapshot_storage.iter().all(|storage_entry| !full_snapshot_storage.contains_key(storage_entry.0)),
+    "There must not be any overlap in the slots of storages between the full snapshot and the incremental snapshot");
```
Should this be a hard assert or just another error result?

carllin replied:
For instance could a malicious validator make a full snapshot for `X` with slots `Y > X` such that those extra slots `Y` collide with slots in future incremental snapshots?

To prevent this, should we also 
1) verify that a full snapshot for `X` only contains slots `<=X`, 
2) and then if we know incremental snapshots based on `X` only contain slots `> X` from this earlier check here: https://github.com/solana-labs/solana/pull/18504/files#r668284384

Then this non-overlapping naturally falls out of the two guarantees

brooksprumo replied:
Ooo good points. I like this idea and will do.

brooksprumo replied:
I've changed the assert to return an error instead. Is that sufficient for now?

I originally put in the assert from your comment here: https://github.com/solana-labs/solana/pull/17875#discussion_r652475517

carllin replied:
Man, what a flip flopper - can't trust this guy with anything 😛

carllin commented on runtime/src/snapshot_utils.rs:
```diff
@@ -649,64 +896,280 @@ pub fn bank_from_snapshot_archive<P>(
+fn check_are_snapshots_compatible<P, Q>(
+    full_snapshot_archive_path: P,
+    incremental_snapshot_archive_path: Q,
+) -> Result<(Slot, Slot)>
```
Since it seems the tuple from this check is never actually used, we could either:

1) just return a bool and have the caller return the `SnapshotError::IncompatibleSnapshots`

2) Return a `Result<()>`

brooksprumo replied:
Ah good catch! This was leftover when I used the return values to set the last full snapshot slot and last incremental snapshot slot. I'll remove those here.

I think I may need them again when doing the `clean_accounts()` work, but that'll be easy to re-add then.

brooksprumo replied:
Done!

carllin commented on runtime/src/serde_snapshot.rs:
```diff
@@ -321,19 +351,64 @@ where
+        ));
+
+    // filter out incremental snapshot storages with slot <= full snapshot slot
+    let incremental_snapshot_storage = snapshot_accounts_db_fields
+        .incremental_snapshot_accounts_db_fields
+        .map_or_else(HashMap::new, |fields| {
+            fields
+                .0
+                .into_iter()
+                .filter(|(slot, _)| *slot > full_snapshot_slot)
+                .collect::<HashMap<_, _>>()
+        });
```
Should we filter here or just return an error? Since we checked earlier via `check_are_snapshots_compatible` that the incremental snapshot was based on this full snapshot, there should be no slots `<= full snapshot slot` right?

brooksprumo replied:
Yeah... this one confused me too. So when AccountsHashVerifier is taking hashes, it's doing it from snapshots. Those snapshots (not the whole snapshot archive) need to have all the storages to compute the hash correctly.

When I was writing tests in `core/tests/snapshots.rs`, which calls functions a little differently to create snapshot archives, I was running into issues where rebuild would fail due to AppendVec ID mismatch for slots less than the full snapshot slot in the incremental snapshot. The accounts_db_fields.0 also has info for all the slots here, so this filtering needs to happen when rebuilding with an incremental snapshot.

Both the accounts_db_fields.0 _and_ the actual storages need to be filtered out. The storages can get filtered out when creating the incremental snapshot archive, and then the accounts_db_fields.0 needs to be filtered out when rebuilding. Here's some more info when I was asking about it on Discord [1].

[1] https://discord.com/channels/428295358100013066/439194979856809985/854856681082978304

carllin replied:
Little confused here, let's have a brief chat to go over it.

brooksprumo replied:
After talking with @carllin, I've created a new issue to track this: #18813. The gist is during serialization of accounts_db for an incremental snapshot, to filter _out_ the storage entries for all slots less-than-or-equal-to the full snapshot slot.

carllin replied:
Thanks!

carllin commented on runtime/src/snapshot_utils.rs:
```diff
@@ -417,9 +543,9 @@ pub fn archive_snapshot_package(
 }
 
-pub fn get_snapshot_paths<P: AsRef<Path>>(snapshot_path: P) -> Vec<SlotSnapshotPaths>
+pub fn get_snapshot_paths<P>(snapshot_path: P) -> Vec<SlotSnapshotPaths>
```
nit: This is getting the storage entry paths right? How about `get_slot_storage_paths`

carllin commented on runtime/src/snapshot_utils.rs:
```diff
@@ -169,7 +209,8 @@ impl Ord for SlotSnapshotPaths {
 
-pub fn package_snapshot<P: AsRef<Path>, Q: AsRef<Path>>(
+/// Package up slot snapshot files, snapshot storages, and slot deltas for a full snapshot.
+pub fn package_snapshot<P, Q>(
```
nit: package_snapshot -> package_full_snapshot

brooksprumo replied:
Yah, I wasn't sure what to do; either rename all the existing things to `full_`, or not. Since currently a "snapshot" is a full snapshot, that's why I didn't add `full_` to all the functions/variables. Inside all the incremental snapshot functions, that's where everything is either `full_` or `incremental_`.

I also wasn't sure how far to take the renaming. Like should it cause gossip/RPC to change names too? I'm not sure on that part. What do you think @carllin ?

brooksprumo replied:
Oh, and also the on-disk snapshots right now are also named "snapshot" and are full snapshots. To handle renaming the on-disk snapshots, I'd need to add code to support both `full_snapshot` and `snapshot` files/directories for a while, for backwards compatibility.

ryoqun replied:
I think existing identifiers should be prefixed with `full_` to avoid any ambiguity.

and file system and rpc / gossip should remain same.

in other words, `full_` rename should occur in all code where appropriate for humans, while keeping compatibility.

it's a bit daunting for me. but could you do that once for all for us, considering your pr's high quality of plumbing?

brooksprumo replied:
@ryoqun Yes, will do.

brooksprumo replied:
OK, done!

carllin replied:
> in other words, full_ rename should occur in all code where appropriate for humans, while keeping compatibility.

Sorry late to the party here, agreed, especially given we're introducing `incremental_snapshots`

carllin commented on runtime/src/snapshot_utils.rs:
```diff
@@ -179,19 +220,104 @@ pub fn package_snapshot<P: AsRef<Path>, Q: AsRef<Path>>(
+    Q: AsRef<Path>,
+{
+    info!(
+        "Package snapshot for bank: {} has {} account storage entries",
```
nit: `Package snapshot -> Package full snapshot`

brooksprumo replied:
Same naming questions from here [1] on how to name snapshots.

[1] https://github.com/solana-labs/solana/pull/18504#discussion_r668915990

carllin replied:
I think the "full_" vs "incremental_" is good for clarity, thanks!

carllin commented on runtime/src/snapshot_utils.rs:
```diff
@@ -179,19 +220,104 @@ pub fn package_snapshot<P: AsRef<Path>, Q: AsRef<Path>>(
+        .tempdir_in(snapshot_path)?;
+
+    do_package_snapshot(
+        bank,
+        snapshot_files,
+        status_cache_slot_deltas,
+        snapshot_package_output_path,
+        snapshot_storages,
+        archive_format,
+        snapshot_version,
+        hash_for_testing,
```
Since all of these arguments are shared between `do_package_snapshot`, `package_incremental_snapshot`, and `package_snapshot`, it might be cleaner to wrap these up in an argument struct

brooksprumo replied:
While implementing this, I ended up feeling like I didn't actually help anything, just move around where I was writing all the parameters, and even adding a bit of work to create the struct and then destructure it later. How strongly would you like this change to go in?

carllin replied:
Heh, looking at it again, I think `package_incremental_snapshot` and  `package_snapshot` may just be better off as constructors for `AccountsPackagePre`, where logic like https://github.com/solana-labs/solana/pull/18504/files#diff-c957621ae265a12e03f7b9c5656e7aa74137223e603cc3cfcc2828dda1af5479R398-R405 could be moved into a constructor.

Also has the added benefit of offloading from the `snapshot_utils` module as it's growing rather large.

brooksprumo replied:
After talking with @carllin, these changes will not be part of this PR. I'll likely pull it into another change later. Basically just pulling the TempDir into constructors for AccountsPackagePre.

brooksprumo replied:
New issue created to track this work: #18815

carllin commented on runtime/src/snapshot_utils.rs:
```diff
@@ -417,9 +543,9 @@ pub fn archive_snapshot_package(
 }
 
-pub fn get_snapshot_paths<P: AsRef<Path>>(snapshot_path: P) -> Vec<SlotSnapshotPaths>
+pub fn get_snapshot_paths<P>(snapshot_path: P) -> Vec<SlotSnapshotPaths>
```
nit: `get_snapshot_paths -> get_slot_snapshot_paths`, 

and a comment along the lines of, "returns a vector of paths to existing snapshot directories under `snapshot_path`"

brooksprumo replied:
omg yes. I wanted to do this myself too. I initially tried to change as little as possible for this PR. Happy to make this change!

brooksprumo replied:
Done!

carllin replied:
nice, much more readable, thanks :)

carllin commented on runtime/src/snapshot_utils.rs:
```diff
@@ -135,6 +161,14 @@ pub struct SlotSnapshotPaths {
     pub snapshot_file_path: PathBuf,
 }
```
really want to rename this to `SlotSnapshotPath`, the plural is confusing here when its only signifying one slot

brooksprumo replied:
I was very confused when I first read through the code. Happy to change this :)

brooksprumo replied:
Done!

carllin replied:
heh, awesome, thanks!

carllin commented on runtime/src/snapshot_utils.rs:
```diff
@@ -1131,9 +1790,82 @@ pub fn bank_to_snapshot_archive<P: AsRef<Path>, Q: AsRef<Path>>(
+    let temp_dir = tempfile::tempdir_in(snapshot_path)?;
+
+    let storages = bank.get_incremental_snapshot_storages(full_snapshot_slot);
+    let slot_snapshot_paths = add_snapshot(&temp_dir, bank, &storages, snapshot_version)?;
```
nit: `slot_snapshot_paths -> slot_snapshot_path`, since it's only one path?

brooksprumo replied:
Done!

carllin commented on runtime/src/snapshot_utils.rs:
```diff
@@ -179,19 +220,104 @@ pub fn package_snapshot<P: AsRef<Path>, Q: AsRef<Path>>(
+fn package_incremental_snapshot<P, Q>(
+    bank: &Bank,
+    incremental_snapshot_base_slot: Slot,
+    snapshot_files: &SlotSnapshotPaths,
```
`snapshot_files -> slot_snapshot_path` since it's only one slot?

brooksprumo replied:
Done!

lijunwangs commented on runtime/src/snapshot_utils.rs:
```diff
@@ -1131,9 +1790,82 @@ pub fn bank_to_snapshot_archive<P: AsRef<Path>, Q: AsRef<Path>>(
+) -> Result<PathBuf> {
+    let snapshot_version = snapshot_version.unwrap_or_default();
+
+    assert!(bank.is_complete());
```
Maybe we can put this common code into a separate function to share with bank_to_snapshot_archive?

brooksprumo replied:
Yeah, I thought about doing that too. Unfortunately the only real common code are the first few lines of `bank.xxx()`, then afterwards they are different depending on full snapshot vs incremental snapshot. I decided to leave them separate for that reason, as the duplication wasn't much. Also, I am anticipating needing to change the impl in the incremental snapshot side w.r.t. clean_accounts(), so that's another reason I've left them separate. Is that OK for now?

lijunwangs commented:
What is the impact for existing validators downloading snapshots from validators generating incremental snapshots? Will they work off the full snapshot available?

brooksprumo commented on runtime/src/snapshot_utils.rs:
```diff
@@ -649,64 +896,280 @@ pub fn bank_from_snapshot_archive<P>(
+        snapshot_path,
+        TMP_SNAPSHOT_PREFIX,
+        &full_snapshot_archive_path,
+        "snapshot untar",
```
Yeah, I wasn't sure how much to rename. See this comment: https://github.com/solana-labs/solana/pull/18504#discussion_r668915990

My thought for this PR was everything that didn't include "incremental" meant it was a full snapshot, which is the way it works today, since all snapshots are full snapshots. Then within the context of incremental snapshot code, I would always add "full" to differentiate. Whatcha think?

brooksprumo commented on runtime/src/snapshot_utils.rs:
```diff
@@ -649,64 +896,280 @@ pub fn bank_from_snapshot_archive<P>(
+        rebuild_bank_from_snapshots_us: measure_rebuild.as_us(),
+        untar_us: full_snapshot_measure_untar.as_us()
+            + incremental_snapshot_measure_untar.map_or(0, |measure| measure.as_us()),
+        verify_snapshot_bank_us: measure_verify.as_us(),
```
Good idea, will do!

brooksprumo replied:
Done!

brooksprumo commented on runtime/src/snapshot_utils.rs:
```diff
@@ -649,64 +896,280 @@ pub fn bank_from_snapshot_archive<P>(
+{
+    fn path_to_filename(path: &Path) -> Result<&str> {
+        path.file_name()
+            .ok_or(SnapshotError::PathParseError("Could not get file name!"))?
```
Good point. I will fix this.

brooksprumo replied:
Done!

brooksprumo commented on runtime/src/snapshot_utils.rs:
```diff
@@ -796,6 +1296,69 @@ fn sort_snapshot_archives(snapshot_archives: &mut Vec<SnapshotArchiveInfo>) {
+{
+    match fs::read_dir(&snapshot_output_dir) {
+        Err(err) => {
+            info!("Unable to read snapshot directory: {}", err);
```
I believe so. I will fix this.

brooksprumo replied:
Done!

brooksprumo commented:
> What is the impact for existing validators downloading snapshots from validators generating incremental snapshots? Will they work off the full snapshot available?

Sorry I could reply to your comment directly, @lijunwangs. Firstly, in this PR there are no changes to the actual way validators function; no incremental snapshots are taken. Going forward, yes, validators that are running older versions without incremental snapshot support will continue to operate the same as today: asking for (and working off) full snapshots.

brooksprumo commented on runtime/src/snapshot_utils.rs:
```diff
@@ -649,64 +896,280 @@ pub fn bank_from_snapshot_archive<P>(
+            incremental_snapshot_unpack_dir,
+            incremental_snapshot_unpacked_snapshots_dir,
+            incremental_snapshot_unpacked_append_vec_map,
+            incremental_snapshot_version,
```
I wondered about this same thing. In a previous PR I had a TODO in the code and Carl answered the question:

https://github.com/solana-labs/solana/pull/17875#discussion_r652452533

So AFAIK the handling here of versions is sufficient. Is that OK? Would you like to see something else?

brooksprumo commented:
@carllin @lijunwangs OK! I had fixed or replied to all the review comments. Please take another pass and let me know what you think. Thanks for all the help so far!

@ryoqun Sorry, I should've added you earlier as well!

ryoqun commented on ledger/src/blockstore_processor.rs:
```diff
@@ -559,7 +559,16 @@ fn do_process_blockstore_from_root(
         ("slot", bank_forks.root(), i64),
         ("forks", initial_forks.len(), i64),
         ("calculate_capitalization_us", time_cap.as_us(), i64),
-        ("untar_us", timings.untar_us, i64),
```
unintentional remove due to merge conflict?

ryoqun replied:
oops. never mind.. prefixed...

ryoqun commented on runtime/src/serde_snapshot.rs:
```diff
@@ -326,19 +356,68 @@ where
+    let (snapshot_version, snapshot_slot, snapshot_bank_hash_info) = snapshot_accounts_db_fields
+        .incremental_snapshot_accounts_db_fields
+        .as_ref()
+        .map(
+            |AccountsDbFields(
+                _,
+                incremental_snapshot_version,
+                incremental_snapshot_slot,
+                incremental_snapshot_bank_hash_info,
+            )| {
+                (
+                    *incremental_snapshot_version,
+                    *incremental_snapshot_slot,
+                    incremental_snapshot_bank_hash_info.clone(),
+                )
+            },
+        )
+        .unwrap_or((
+            full_snapshot_version,
+            full_snapshot_slot,
+            full_snapshot_bank_hash_info,
+        ));
+
+    // filter out incremental snapshot storages with slot <= full snapshot slot
+    let incremental_snapshot_storage = snapshot_accounts_db_fields
+        .incremental_snapshot_accounts_db_fields
+        .map_or_else(HashMap::new, |fields| {
+            fields
+                .0
+                .into_iter()
+                .filter(|(slot, _)| *slot > full_snapshot_slot)
+                .collect::<HashMap<_, _>>()
+        });
+
+    // There must not be any overlap in the slots of storages between the full snapshot and the incremental snapshot
+    incremental_snapshot_storage
+        .iter()
+        .all(|storage_entry| !full_snapshot_storage.contains_key(storage_entry.0)).then(|| ()).ok_or_else(|| {
+            io::Error::new(io::ErrorKind::InvalidData, "Snapshots are incompatible: There are storages for the same slot in both the full snapshot and the incremental snapshot!")
+        })?;
```
how about extracting this code block to a newly-defined method at `SnapshotAccountsDbFields`?

ryoqun replied:
i think this incremental/full snapshot new logic can be well encapsulated, considering this following code (https://github.com/solana-labs/solana/pull/18504/files#r671230020) like this:

```rust

impl SnapshotAccountsDbFields {
    fn combined_storages(...) -> Vec<_> {
    }
}

```

brooksprumo replied:
Done!

ryoqun commented on runtime/src/serde_snapshot.rs:
```diff
@@ -326,19 +356,68 @@ where
     }
 
-    let storage = storage.into_iter().collect::<Vec<_>>();
+    let storage = full_snapshot_storage
+        .into_iter()
+        .chain(incremental_snapshot_storage.into_iter())
+        .collect::<Vec<_>>();
```
this

ryoqun commented on runtime/src/snapshot_utils.rs:
```diff
@@ -417,43 +543,56 @@ pub fn archive_snapshot_package(
+                entry.ok().and_then(|e| {
+                    e.path()
+                        .file_name()
+                        .and_then(|n| n.to_str().map(|s| s.parse::<u64>().ok()))
```
nits: `u64` => `Slot`

brooksprumo replied:
Done!

ryoqun commented on runtime/src/snapshot_utils.rs:
```diff
@@ -489,7 +635,7 @@ where
     data_file_stream.flush()?;
 
-    let consumed_size = data_file_stream.seek(SeekFrom::Current(0))?;
+    let consumed_size = data_file_stream.stream_position()?;
```
TIL :)

ryoqun commented on runtime/src/snapshot_utils.rs:
```diff
@@ -649,66 +901,188 @@ pub fn bank_from_snapshot_archive<P>(
 ) -> Result<(Bank, BankFromArchiveTimings)>
 where
     P: AsRef<Path> + std::marker::Sync,
+    Q: AsRef<Path> + std::marker::Sync,
```
dumb question: so, `full_snapshot_archive_path` and `incremental_snapshot_archive_path` could be different types for any given this fn invocation? if possible, I want to avoid another type param (Q in this case)...

brooksprumo replied:
In theory yes the types could be different, but in practice probably not. I forget which Rust resource I read that recommended this, but I did see it someone, so that one could be a `&Path` and the other could be a `PathBuf` and the function doesn't need to care about that. Without `Q`, then both FSS and ISS paths need to be the same type. Would you like me to remove `Q` here?

ryoqun replied:
@brooksprumo yeah, in general, fewer type params are desired to make code simpler if our current code base allows.

> I read that recommended this, but I did see it somewhere... So that one could be a &Path and the other could be a PathBuf and the function doesn't need to care about that.

seems you have good memory. I think this is perfectly good practice for library context. Unless sdk or related, our codebase is mostly application. so, we don't need to be that _generic_ :)

brooksprumo replied:
Done!

ryoqun commented on runtime/src/snapshot_utils.rs:
```diff
@@ -649,66 +901,188 @@ pub fn bank_from_snapshot_archive<P>(
-        &snapshot_tar,
-        unpack_dir.as_ref(),
+
+    let full_snapshot_unarchive_preparation_result = do_snapshot_unarchive_preparation(
```
nits: rename to `prepared_snapshot_archive`? a bit mouthful variable name, imo... like `Option<_>`, `Result<_>` type variable doesn't need to be suffixed with `_result`.

brooksprumo replied:
I like it. Done!

ryoqun commented on runtime/src/snapshot_utils.rs:
```diff
@@ -154,53 +194,139 @@ pub enum SnapshotError {
+    ParseSnapshotArchiveFileNameError(String),
+
+    #[error("snapshots are incompatible: full snapshot slot ({0}) and incremental snapshot base slot ({1}) do not match")]
+    IncompatibleSnapshots(Slot, Slot),
```
nits: `MismatchedBaseSlot`? `Incompatible` a bit sounds too general.

brooksprumo replied:
Done!

ryoqun commented on runtime/src/snapshot_utils.rs:
```diff
@@ -750,14 +1143,36 @@ fn parse_snapshot_archive_filename(archive_filename: &str) -> Option<(Slot, Hash
+    let regex = Regex::new(INCREMENTAL_SNAPSHOT_ARCHIVE_FILENAME_REGEX);
+
+    regex.ok()?.captures(archive_filename).and_then(|captures| {
+        let base_slot = captures.get(1).map(|x| x.as_str().parse::<Slot>())?.ok()?;
```
maybe, there is no `captures.iter()` or something similar? I like multiple `capture_iter.next()`s than `.get(1)`/`.get(2)`/`.get(3)`.... (no magic numbers if possible..)

brooksprumo replied:
I agree; no magic numbers if possible! I'll take a look to see if there's another way (i.e. `.iter()`) to accomplish this.

brooksprumo replied:
Done!

ryoqun commented on runtime/src/snapshot_utils.rs:
```diff
@@ -798,6 +1213,75 @@ fn sort_snapshot_archives(snapshot_archives: &mut Vec<SnapshotArchiveInfo>) {
+fn sort_incremental_snapshot_archives(
+    incremental_snapshot_archives: &mut Vec<IncrementalSnapshotArchiveInfo>,
+) {
+    incremental_snapshot_archives.sort_unstable_by(|a, b| {
```
how about defining `Ord` for `IncrementalSnapshotArchiveInfo`, so that we can just call `sort()`?

brooksprumo replied:
I actually purposely didn't want to do that, since I didn't feel there was a _universal_ ordering of this type that applied in all situations. We know in the context of the _callers_ of this function that they want the sorting to be first by base slot in descending order, then by "regular" slot in descending order, but maybe that's not the case for the next caller to come along.

What do you think? I can definitely change this to use `Ord` instead, as we know all the callers of this function.

ryoqun replied:
ooh, `a` and `b` is swapped in the code... 

could you at least comment it. I think it's too easy to miss this...

or better, I rather strive to keep to avoid descending order except performance sensitive code.

in this case, I'd write like this:

```rust
impl Ord for IncrementalSnapshotArchiveInfo {
   ... // COMMENT: yea, (base slot asc, regular slot asc) should be kind of universal order, assuming it's aligning with chronological order in the  snapshot lifecycle.
}

...


sort_incremental_snapshot_archives(&mut incremental_snapshot_archives); // COMMENT: or just call straightly: incremental_snapshot_archives.sort_unstable();


// pick the highest (i.e. latest one) one
incremental_snapshot_archives.into_iter().rev().next()
```

brooksprumo replied:
OK, I've put `Ord` back in and reworked the code to use `sort_unstable()` directly.

ryoqun commented on runtime/src/snapshot_utils.rs:
```diff
@@ -536,18 +679,121 @@ where
+
+/// Before running the deserializer function, perform common operations on the snapshot archive
+/// files, such as checking the file size and opening the file into a stream.
+fn prepare_deserialize_snapshot_data_file<P>(
```
nits: naming: `create_snapshot_data_file_stream`? `prepare_` is a bit too general. I know you picked `prepare_` because of the extra checks to encompass these functionalities. but as par defensive programming, doing routine checks are obvious unless we're too lazy. xD

brooksprumo replied:
Goo points. Done!

ryoqun commented on runtime/src/snapshot_utils.rs:
```diff
@@ -817,23 +1311,67 @@ where
+    // snapshots, only a subset of the incremental snapshots for the lastest full snapshot could be
+    // kept.  This could reuse maximum_snapshots_to_retain, or use a new field just for incremental
+    // snapshots.
+    let last_full_snapshot_slot = get_highest_snapshot_archive_slot(snapshot_output_dir.as_ref());
```
if last_full_snapshot_slot is None, purging interval snapshots should happen? I don't think `Some(archive_info.base_slot) < None` evaluates to `true`...

brooksprumo replied:
If `last_full_snapshot_slot` is `None`, then there are also no incremental snapshots to remove (since it is not possible to create an incremental snapshot until a full snapshot is created first). I think this should work OK then, yes?

ryoqun replied:
@brooksprumo heh, welcome to happy fs programming especially under byzantine (or untrained ops) environment... :) well, admins can manually remove all of full snapshots and then `solana-validator` will leave these outdated and bricked incremental snapshots indefinitely with this current code? I want to prevent the garbage from piling to avoid another mysterious support ticket.. lol

brooksprumo replied:
Gotcha! I've made it so now if there are no full snapshots that all the incremental snapshots will be purged. Done!

ryoqun commented on runtime/src/snapshot_utils.rs:
```diff
@@ -798,6 +1213,75 @@ fn sort_snapshot_archives(snapshot_archives: &mut Vec<SnapshotArchiveInfo>) {
+                if let Ok(entry) = entry {
+                    let path = entry.path();
+                    if path.is_file() {
+                        if let Some((base_slot, slot, hash, archive_format)) =
+                            parse_incremental_snapshot_archive_filename(
+                                path.file_name().unwrap().to_str().unwrap(),
+                            )
+                        {
+                            return Some(IncrementalSnapshotArchiveInfo {
+                                inner: SnapshotArchiveInfo {
+                                    path,
+                                    slot,
+                                    hash,
+                                    archive_format,
+                                },
+                                base_slot,
+                            });
+                        }
```
maybe extract to `IncrementalSnapshotArchiveInfo::new_from_path(path) -> Result<Self>`?

brooksprumo replied:
Done!

ryoqun commented on runtime/src/snapshot_utils.rs:
```diff
@@ -1038,14 +1619,20 @@ pub fn verify_snapshot_archive<P, Q, R>(
-pub fn purge_old_snapshots(snapshot_path: &Path) {
-    // Remove outdated snapshots
-    let slot_snapshot_paths = get_snapshot_paths(snapshot_path);
+/// Remove outdataed snapshots
```
nits: `outdated`

brooksprumo replied:
Done!

ryoqun commented:
> @ryoqun Sorry, I should've added you earlier as well!

no worries. i was lurking. xD late for the party but i finally reviewed _the great of plumbing_ pr. well cooked and documented work. :)

brooksprumo commented:
@ryoqun OK! I think I've addressed all of your review comments. Most are fixed in the code now. I replied to the other comments with questions for clarification or how to proceed. I'll re-request the review as well. Thanks!

ryoqun commented on runtime/src/snapshot_utils.rs:
```diff
@@ -417,43 +574,56 @@ pub fn archive_snapshot_package(
+            })
+            .map(|slot| {
+                let snapshot_file_name = get_snapshot_file_name(slot);
+                // Yep, you read that right, the same snapshot file name is joined twice in a row.
+                // "Why" you ask?  This I do not know.
```
i think this is just overlook from original implementation... so could you word like this?

```
// The redundant snapshot_file_name in path is unintentional and should be simplified
// kept just for compatibility.
```

the motto here is that no Qs into our own code... xD

brooksprumo replied:
Done!

ryoqun commented on runtime/src/snapshot_utils.rs:
```diff
@@ -649,66 +934,177 @@ pub fn bank_from_snapshot_archive<P>(
+    account_paths: &[PathBuf],
+    archive_format: ArchiveFormat,
+    parallel_divisions: usize,
+) -> Result<UnarchivePreparationResult>
```
> `Result<UnarchivePreparationResult>`

I think the type sig of this fn is a bit mouthful and I want to avoid 2 nested `Result`s... could you reword this function like this?

```rust
struct UnarchivedSnapshot { // <= renamed from UnarchivePreparationResult
}


fn unarchive_snapshot(...) -> Result<UnarchivedSnapshot> { // renamed from do_snapshot_unarchive_preparation
}
```

better names are welcome. btw, I don't like too reliance of `MetaFooBar`, `FooBarInfo`, `FooBarContext` across our codebase, so I'm eager to remove them. xD

i think enginner's job is to name things properly with good vocabulary for future generation's code readability.. lol

brooksprumo replied:
I like this, done!

ryoqun commented on runtime/src/serde_snapshot.rs:
```diff
@@ -326,25 +397,31 @@ where
             .unwrap_or_else(|err| panic!("Failed to create directory {}: {}", path.display(), err));
     }
 
-    let storage = storage.into_iter().collect::<Vec<_>>();
```
good bye bad singular name. :)

ryoqun commented on runtime/src/snapshot_utils.rs:
```diff
@@ -500,14 +679,11 @@ where
 }
 
-fn deserialize_snapshot_data_file_capped<F, T>(
+fn deserialize_snapshot_data_file_capped<T: Sized>(
```
out of curious, why `T: Sized` trait bound is needed now?

ryoqun replied:
oh i see. yeah, `impl Trait` requires `Sized`, iirc. no problem then. :)

brooksprumo commented:
@lijunwangs @ryoqun @carllin Checks have passed, and all review comments have been addressed. I've replied with answers/questions to any outstanding review comments. PR is ready for another round of reviewing. Thanks in advance!

ryoqun commented:
> - [ ] **Assert vs error when storages overlap.** [Thread 1 here](https://github.com/solana-labs/solana/pull/18504#discussion_r668281871), and [thread 2 here](https://github.com/solana-labs/solana/pull/18504#discussion_r668284384)

@brooksprumo as a context, we want to support untrusted snapshots eventually in the (distant) future. so, panic! shouldn't be reachable when consuming (= untaring) any downloaded snapshots. #7167

That context might help to decide panic!/error?

Also, well-groomed pr description. :)

brooksprumo commented:
> @brooksprumo as a context, we want to support untrusted snapshots eventually in the (distant) future. so, panic! shouldn't be reachable when consuming (= untaring) any downloaded snapshots. #7167
> 
> That context might help to decide panic!/error?

@ryoqun Yes, thank you. Makes sense!

> Also, well-groomed pr description. :)

Thanks :)

carllin commented on runtime/src/serde_snapshot.rs:
```diff
@@ -74,6 +74,62 @@ struct AccountsDbFields<T>(
+                incremental_snapshot_slot,
+                incremental_snapshot_bank_hash_info,
+            )) => {
+                let full_snapshot_storages = self.full_snapshot_accounts_db_fields.0;
+                let full_snapshot_slot = self.full_snapshot_accounts_db_fields.2;
```
heh, really wish we'd gone with named fields for `AccountsDbFields` 😛

brooksprumo replied:
Me too. I had grabbed `.1` for `slot` and all my tests were failing until I realized...

carllin commented on runtime/src/serde_snapshot.rs:
```diff
@@ -74,6 +74,62 @@ struct AccountsDbFields<T>(
+    /// incremental snapshot, this returns the AccountsDbFields from the full snapshot.  Otherwise
+    /// this uses the version, slot, and bank hash info from the incremental snapshot, then the
+    /// combination of the storages from both the full and incremental snapshots.
+    fn collapse_into(self) -> Result<AccountsDbFields<T>, Error> {
+        match self.incremental_snapshot_accounts_db_fields {
+            None => Ok(self.full_snapshot_accounts_db_fields),
+            Some(AccountsDbFields(
+                mut incremental_snapshot_storages,
+                incremental_snapshot_version,
+                incremental_snapshot_slot,
+                incremental_snapshot_bank_hash_info,
+            )) => {
+                let full_snapshot_storages = self.full_snapshot_accounts_db_fields.0;
+                let full_snapshot_slot = self.full_snapshot_accounts_db_fields.2;
+
+                // filter out incremental snapshot storages with slot <= full snapshot slot
+                incremental_snapshot_storages.retain(|slot, _| *slot > full_snapshot_slot);
+
+                // There must not be any overlap in the slots of storages between the full snapshot and the incremental snapshot
+                incremental_snapshot_storages
+                    .iter()
+                    .all(|storage_entry| !full_snapshot_storages.contains_key(storage_entry.0)).then(|| ()).ok_or_else(|| {
+                        io::Error::new(io::ErrorKind::InvalidData, "Snapshots are incompatible: There are storages for the same slot in both the full snapshot and the incremental snapshot!")
+                    })?;
+
+                let mut combined_storages = full_snapshot_storages;
+                combined_storages.extend(incremental_snapshot_storages.into_iter());
+
+                Ok(AccountsDbFields(
+                    combined_storages,
+                    incremental_snapshot_version,
+                    incremental_snapshot_slot,
+                    incremental_snapshot_bank_hash_info,
+                ))
+            }
+        }
+    }
+}
+
```
nice, like what you did here

lijunwangs reviewed (approved):
Looks good to me.

brooksprumo commented:
@carllin @lijunwangs I had to fix a merge conflict, so your approvals were removed. Nothing else changed, but let me know if you'd like me to wait for you to re-review before I merge.

@ryoqun Would you like to re-review before I merge?

ryoqun commented on runtime/src/snapshot_utils.rs:
```diff
@@ -750,105 +1224,220 @@ fn archive_format_from_str(archive_format: &str) -> Option<ArchiveFormat> {
+    // In case there are incremental snapshots but no full snapshots, make sure all the incremental
+    // snapshots are purged.
+    let last_full_snapshot_slot =
+        get_highest_full_snapshot_archive_slot(&snapshot_archives_dir).unwrap_or(Slot::MAX);
```
:eyes: good

ryoqun commented on runtime/src/snapshot_utils.rs:
```diff
@@ -665,81 +1011,209 @@ pub fn bank_from_snapshot_archive<P>(
+
+    let unpacked_version_file = unpack_dir.path().join("version");
+    let mut snapshot_version = String::new();
+    File::open(unpacked_version_file).and_then(|mut f| f.read_to_string(&mut snapshot_version))?;
```
here

ryoqun replied:
also, I just noticed we're better off checking file size of `snapshot_version`? for eventual untrusted snapshot thing?

Otherwise, it could be dos vector if crafted snapshot contains a 100G zeroed snapshot_version file because we read all to memory first....

unrelated to this pr off course. so could be done as a follow up pr...

ryoqun replied:
ref: #7167

brooksprumo replied:
New issue created: #18841

ryoqun commented on runtime/src/snapshot_utils.rs:
```diff
@@ -925,72 +1514,116 @@ fn untar_snapshot_in<P: AsRef<Path>>(
+fn verify_unpacked_snapshots_dir_and_version(
+    unpacked_snapshots_dir_and_version: &UnpackedSnapshotsDirAndVersion,
+) -> Result<(SnapshotVersion, BankSnapshotInfo)> {
+    let snapshot_version_str = unpacked_snapshots_dir_and_version.snapshot_version.trim();
```
maybe `.trim()` could be moved here?: https://github.com/solana-labs/solana/pull/18504/files#r674890169

That's to group similar code together.

ryoqun replied:
also, sanitizing untrusted input early in the codepath is always better practice. :)

brooksprumo replied:
Done!

ryoqun reviewed (approved):
I'd say a masterpiece of plumbing pr ever. xD

I only left some minor comments. merge anytime without needing another review round.
